### ISSUE TITLE
Easier way to build menus

### ISSUE THREAD

**Issue by jenstornell** [jenstornell/kirby-plugin-wishlist#25]
>Easier way to build menus without going through several each loops. @jimbobrjames

**Comment by krisak** [comment]
So are we talking about some sort of tree sturcture walker? Or panel field that enables to make menus more easily?

**Comment by jenstornell** [comment]
@krisak I've sent the question to @jimbobrjames which is the author of this idea.

**Comment by CodeStalker** [comment]
Hey thanks, I just came here to put my reply but you beat me to it :)

**Comment by krisak** [comment]
I can see potential of that. I don't need it right now but i know when this would be very useful. 

Not that i am fan of Wordpress but just for reference and inspiration - [](https://codex.wordpress.org/Class_Reference/Walker )

It can grow to pretty big thing. So i think finding the right kirby zen balance can be a challange.

**Comment by CodeStalker** [comment]
I can't stand Wordpress. I can proudly say i've never built a clients site with it :)

I think a framework style approach is probably the way to go, rather then a tightly closed plugin that could end up being restrictive... if there were some extra hooks that allow things like this to built more easily, that can only be a good thing. I think it has potential beyond site menus - it could be used to build custom RSS feeds and XML sitemaps.

I've been using Kirby for about 6 months now after using Textpattern for years, so forgive me if ive mis understood something but from what i gather, when you do `$page->children` it only goes one level down. If this was capable of recursively finding all decendents and knowing when one of those pages is a parent in its self, that would be enough.

**Comment by CodeStalker** [comment]
Maybe it would be more beneficial to extend Kirby's templates. One thing I really miss from Textpattern is its [template language](https://docs.textpattern.io/tags/). Its dead simple, and looks like HTML. I think some inspiration could be taken from there.

Is there an equivelent in Kirby to Textpatterns "different" tag. This could be used to build menus. [different tag](https://docs.textpattern.io/tags/if_different).

**Comment by jenstornell** [comment]
@krisak Personally I think walkers are complicating things and I don't see the benefit of adding another layer of abstraction.

@CodeStalker I'm aware of the problems with infinity depth menus because a foreach loop only loops the current set of items. There are problaby a few different solutions to the problems but one I like is this one:

https://github.com/bastianallgeier/kirbycms-extensions/blob/master/snippets/treemenu/treemenu.php

Basically it's a snippet that calls itself and send the new page object. In the end it means very little code, no new things to learn and a code that is easy to understand.

As a plugin or not I'm not sure. Maybe this is good enough?

https://github.com/bastianallgeier/kirbycms-extensions/tree/master/snippets/treemenu

**Comment by krisak** [comment]
Well i thought you want to have builder of the menu. Not just way to get stuff.

I could see myself using that on sites that are not custom. Meaning things like template that somebody installs and plugin (maybe with some options in panel) figures stuff out and generates menu.


What you are talking about @CodeStalker can be done quite easily with ```$page->depth()```
 and ```$page–>index()```. If you just loop through index() and use depth() and filter() you get everything you need. It probably is not the most efficient code (it would matter with a lot of children) but that's probably negligible.

@CodeStalker about templates - its ongoing debate. Personally i am in agreement with Kirby core devs - its uncool opinion but... PHP already is a templating language. You can totally use templating language with kirby (like twig https://github.com/fvsch/kirby-twig). But if you look even at their example. 
```php
<h1><?= $page->title() ?></h1>

versus

<h1>{{ page.title }}</h1>
```
Its just different syntax. The only thing templates in PHP give you is certain safety, because you are limiting what you can do in the template (you cant just write too much logic in the template). But that just comes down to your morale. Just don't do dirty stuff in templates :).

@CodeStalker btw i would also argue that the textpattern template example is exactly why i like kirby. It might seem easy html to you but i had hard times understanding what it does because it is unlike php or any other templating languages i know.

**Comment by CodeStalker** [comment]
@krisak All fair points. These things often end up dividing people. I'm a designer/developer rather then a computer science/developer. Whilst I get your point, PHP is in itself a template language, I find it hard to read. I tried the recent Blade plugin but that ended up causing issues so I had to revert back to vanilla Kirby. I might try the Twig plugin you mentioned. Or perhaps its time i actually learnt PHP. Maybe I want nice tags through laziness. That said, Textpattern does have a PHP tag that allows you to write raw PHP in a template if you so fancy. Just wrap it in <txp:php> instead of the PHP open tag and your good to go.

@jenstornell @krisak Thanks for the pointers, ill look into those approaches to menus.

**Comment by krisak** [comment]
@CodeStalker I think it might just take some time to get out of "textpattern mindset".

Kirby is extremely easy. I don't know any simpler cms. The moment you realize most of the stuff is just collection with helpful methods and that there are shortcuts like page()... stuff will click.

Btw i hate PHP with all my heart. Seriously its not very good language. But for the usecase it gets stuff done.

**Comment by CodeStalker** [comment]
@krisak The reason I moved to Kirby is because Textpattern's development stalled for a couple of years and they were only releasing security patches rather then new features or evolving the system. This was because they had gone as far as they could with the current codebase and were trying to work out how to best go about a rewrite from the ground up. I got fed up with waiting. I picked Kirby because its not that dissimilar to Textpattern and it clicked with me more then other systems. I will go back to Textpattern one day, its a very good system and I think easier then Kirby to develop on.

Yeah PHP is horrible :)
